### ISSUE TITLE
crypto/Kconfig Broken

### ISSUE THREAD

**Issue by HacKurx** [dapperlinux/dapper-secure-kernel-patchset-stable#3]
Hi,

The following errors detected during compilation of version 4.9.92 : 
```
crypto/Kconfig:15:error: recursive dependency detected!
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
crypto/Kconfig:15:	symbol CRYPTO is selected by BIG_KEYS
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
security/keys/Kconfig:44:	symbol BIG_KEYS depends on CRYPTO_ANSI_CPRNG
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
crypto/Kconfig:1613:	symbol CRYPTO_ANSI_CPRNG depends on CRYPTO
warning: (MAC802154 && MAC80211) selects CRYPTO_CCM which has unmet direct dependencies (CRYPTO)
warning: (MAC80211 && MACSEC) selects CRYPTO_GCM which has unmet direct dependencies (CRYPTO)
warning: (MAC802154 && FS_ENCRYPTION) selects CRYPTO_CTR which has unmet direct dependencies (CRYPTO)
```

Please make a fork of the linux kernel or the minipli unofficial_grsec branch in order to submit commits more easily.

**Comment by HacKurx** [comment]
In "security/keys/Kconfig" remove:
`depends on (CRYPTO_ANSI_CPRNG = y || CRYPTO_DRBG = y)`

**Comment by HacKurx** [comment]
There are some things I don't understand in your file "dapper-secure-kernel-patchset-4.9.92-2018-05-27.patch" like this:

```
diff --git a/security/keys/big_key.c b/security/keys/big_key.c
index e628817..90c0740 100644
--- a/security/keys/big_key.c
+++ b/security/keys/big_key.c
@@ -1,6 +1,5 @@
 /* Large capacity key type
  *
- * Copyright (C) 2017 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
  * Copyright (C) 2013 Red Hat, Inc. All Rights Reserved.
  * Written by David Howells (dhowells@redhat.com)
  *
```

Why it's not indicated in the "Omitted" part ?

This simple example must impact a lot of files, so be careful.

**Comment by miroR** [comment]
Here's how it looks like with my system.

```
scripts/kconfig/conf  --silentoldconfig Kconfig
crypto/Kconfig:15:error: recursive dependency detected!
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
crypto/Kconfig:15:  symbol CRYPTO is selected by BIG_KEYS
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
security/keys/Kconfig:44:   symbol BIG_KEYS depends on CRYPTO_ANSI_CPRNG
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
crypto/Kconfig:1613:    symbol CRYPTO_ANSI_CPRNG depends on CRYPTO
warning: (INET_ESP && INET6_ESP) selects CRYPTO_ECHAINIV which has unmet direct dependencies (CRYPTO)

```

( Just confirming this issue on my part. )

**Comment by msr50** [comment]
So, this all originates from some patches from 4.9.53 which broke things, and I explained it in my commit for 4.9.54, and you can read a full explanation here:
https://github.com/dapperlinux/dapper-secure-kernel-patchset-stable/commit/afc777df6c430b1cce111c162350542570e00a50

Long story short the size_overflow gcc plugin warned me about an overflow in a crypto testsuite, and it was introduced by some patches from upstream. At the time, I didn't know how to fix it, so I decided to revert the patch and investigate the issue when I had time. I haven't had time, and the reverted patch still sits in my tree today.

You can find it here: https://github.com/dapperlinux/dapper-secure-kernel-patchset-stable/blob/master/patches/debug.patch

There was recently a merge conflict with the patch that HacKurx pointed out, which I just merged into my reverted patch and caused the problem.

Now, its not in the "omitted" part because this pre-dates any omitting that I had to do to get 4.9.75+ working. You are right, it is bad form and I should fix it. 

When I have some time, I will remove my "fix" and do the omitting like I have done for other patches. If I get time, I will include the patches and try and debug the size_overflow plugin error. I did have a look at minipli's tree at the time, but it seems he is not effected. The problem is, size_overflow halts the kernel, which makes things unusable until it is debugged and fixed.

**Comment by miroR** [comment]
@HacKurx if you could look up:
https://github.com/dapperlinux/dapper-secure-kernel-patchset-stable/issues/4#issuecomment-393427028
thanks in advance!

**Comment by msr50** [comment]
Okay, so I went and omitted the changes to big_key properly, and have included some patch difference files from 4.9.53.

This fixes the config errors that we get when we go to compile. Now, this does not fix the underlying reason why I omitted these patches in the first place, since if I include the big_key changes, size_overflow plugin halts the kernel on some aead crypto changes.

Again, my writeup is in this commit: https://github.com/dapperlinux/dapper-secure-kernel-patchset-stable/commit/afc777df6c430b1cce111c162350542570e00a50

We will have to go and take another look at that bug another day. Might open an issue for that one specifically.

Anyway, I went and had a look for any more "temporary" omits I introduced into the patchset, and found no more. Its why the patch had a name of patches/debug.patch to remind me to return to it. It now lives in the 4.9.53 omitted patch.
